pkg/controllers/user/multiclusterapp/controller_multiclusterapp.go

```diff
@@ -0,0 +1,335 @@
+package multiclusterapp
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	corev1 "github.com/rancher/types/apis/core/v1"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	pv3 "github.com/rancher/types/apis/project.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+
+	"k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	globalScopeAnswersKey     = "global"
+	creatorIDAnn              = "field.cattle.io/creatorId"
+	multiClusterAppIDSelector = "mcapp"
+	projectIDFieldLabel       = "field.cattle.io/projectId"
+	multiClusterAppPrefix     = "mcapp-"
+	globalNamespace           = "cattle-global-data"
+)
+
+type MCAppController struct {
+	apps                  pv3.AppInterface
+	appLister             pv3.AppLister
+	multiClusterApps      v3.MultiClusterAppInterface
+	multiClusterAppLister v3.MultiClusterAppLister
+	namespaces            corev1.NamespaceInterface
+	templateVersionLister v3.TemplateVersionLister
+	clusterLister         v3.ClusterLister
+	projectLister         v3.ProjectLister
+	clusterName           string
+}
+
+func Register(ctx context.Context, cluster *config.UserContext) {
+	m := &MCAppController{
+		apps:                  cluster.Management.Project.Apps(""),
+		appLister:             cluster.Management.Project.Apps("").Controller().Lister(),
+		namespaces:            cluster.Core.Namespaces(""),
+		multiClusterApps:      cluster.Management.Management.MultiClusterApps(""),
+		multiClusterAppLister: cluster.Management.Management.MultiClusterApps("").Controller().Lister(),
+		clusterLister:         cluster.Management.Management.Clusters("").Controller().Lister(),
+		projectLister:         cluster.Management.Management.Projects("").Controller().Lister(),
+		clusterName:           cluster.ClusterName,
+		templateVersionLister: cluster.Management.Management.TemplateVersions("").Controller().Lister(),
+	}
+	m.multiClusterApps.AddHandler(ctx, "multi-cluster-app-controller", m.sync)
+}
+
+func (m *MCAppController) sync(key string, mcapp *v3.MultiClusterApp) (runtime.Object, error) {
+	if mcapp == nil || mcapp.DeletionTimestamp != nil {
```

Comment:
any chance we can extract some calls to their own methods? the sync method looks too long